api/src/main/java/org/apache/iceberg/types/TypeUtil.java

```diff
@@ -182,11 +181,10 @@ public static Map<Integer, String> indexQuotedNameById(
   }
 
   public static Map<String, Integer> indexByLowerCaseName(Types.StructType struct) {
-    Map<String, Integer> indexByLowerCaseName = Maps.newHashMap();
+    ImmutableMap.Builder<String, Integer> builder = ImmutableMap.builder();
     indexByName(struct)
-        .forEach(
-            (name, integer) -> indexByLowerCaseName.put(name.toLowerCase(Locale.ROOT), integer));
-    return indexByLowerCaseName;
+        .forEach((name, integer) -> builder.put(name.toLowerCase(Locale.ROOT), integer));
+    return builder.buildOrThrow();
```

Comment:
I'll make the change but it kind of assumes that only two fields will ever have conflicting lowercase names. What if the schema contains fields named `data`, `DATA` and `Data`?

(I agree having a schema with both `data` and `DATA` would be rare. Having a schema with `data`, `DATA` and `Data` would be another order of magnitude rare.)

Reply:
I have manually tested this change and the result is below
```
java.lang.IllegalArgumentException: Unable to build field name to id mapping because two fields have the same lower case name: order_date and ORDER_DATE
```